helix-term/src/commands.rs

```diff
@@ -185,7 +185,9 @@ pub fn move_line_end(cx: &mut Context) {
 
         // Line end is pos at the start of next line - 1
         // subtract another 1 because the line ends with \n
-        let pos = text.line_to_char(line + 1).saturating_sub(2);
+        let pos = text
+            .line_to_char(line + 1)
+            .saturating_sub(doc.line_ending().len() + 1);
```

Comment:
I don't think this correct (although not your fault--the previous version wasn't correct either).

1. We don't need the final `+ 1` inside the saturating sub, since that's now accounted for in the line ending length itself.
2. This code assumes that all line endings in the document match the document's line ending setting, which is not guaranteed.  We need to check the actual line ending of the line.
3. The last line in a `Rope` or `RopeSlice` doesn't have a line ending at all, so in that case there should be no subtraction.

I think what we want to do here is create a new function that just detects the line ending of a single line, and then use that here.  I think the function would look like this (untested):

```rust
/// Returns the passed line's line ending, if any.
fn get_line_ending(line: &RopeSlice) -> Option<LineEnding> {
    // Last character as str.
    let g1 = line.slice(line.len_chars().saturating_sub(1)..).as_str().unwrap();

    // Last two characters as str, or empty str if they're not contiguous.
    // It's fine to punt on the non-contiguous case, because Ropey guarantees
    // that CRLF is always contiguous.
    let g2 = line.slice(line.len_chars().saturating_sub(2)..).as_str().unwrap_or("");

    // First check the two-character case for CRLF, then check the single-character case.
    str_to_line_ending(g2).or_else(|_| str_to_line_ending(g1))
}
```

(This function could also be re-used in the document line detecting code.)

Then the code here would look like this (also untested):

```rust
let pos = text
    .line_to_char(line + 1)
    .saturating_sub(get_line_ending(text.line(line)).map(|le| le.len_chars()).unwrap_or(0));
```

Reply:
This requires us to query `doc.line()` first, would `doc.line_end(line_num)` be a better API maybe?

Reply:
Thinking about it, your version can be used in contexts without doc so it's probably better: https://github.com/helix-editor/helix/blob/1c2585202145467f0fde7ad9c571e462081c3656/helix-core/src/movement.rs#L41

Reply:
> would doc.line_end(line_num) be a better API maybe?

I thought about that too.  Having a `Document::line_end(line_num)` would certainly be more efficient.  But I think for the time being something a little more general-purpose makes more sense.  It's an easy case to special-case and optimize later if we want to.